### ISSUE TITLE
differences between nGPT and nTransformers

### ISSUE THREAD

**Issue by rotem154154** [lucidrains/nGPT-pytorch#4]
https://github.com/lucidrains/nGPT-pytorch/blob/cb7185c66197de878ae478f4ec66d82e14c7c9a0/nGPT_pytorch/nGPT.py#L452-L454
https://github.com/lucidrains/nGPT-pytorch/blob/cb7185c66197de878ae478f4ec66d82e14c7c9a0/nGPT_pytorch/nTransformer.py#L414-L420

It seems that there are some key differences between the forward() implementations in nGPT and nTransformer.
The mask parameter is not passed to the attention layer in nGPT, and l2norm is not applied to the outputs.

**Comment by lucidrains** [comment]
@rotem154154 ah, nice catch with the `mask`

the l2norm in `nGPT` is actually absorbed into the `Residual` module for reuse in another project

**Comment by lucidrains** [comment]
@rotem154154 are you seeing something with nGPT?

**Comment by rotem154154** [comment]
I haven’t tried the model yet, just noticed the differences while reviewing the code.

**Comment by lucidrains** [comment]
@rotem154154 sg Rotem, let me know if you see anything interesting, whatever you can say of course

**Comment by rotem154154** [comment]
Ok, I think I found something—not a bug, just a suggestion for improvement.
I tried using parameters similar to GPT-2 124M from Karpathy's llm.c and ended up with 162M. The main difference seems to be the weight-tying:
https://github.com/karpathy/llm.c/blob/7ecd8906afe6ed7a2b2cdb731c042f26d525b820/train_gpt2.py#L140

It could be nice to have a boolean option to enable weight tying :)

**Comment by lucidrains** [comment]
@rotem154154 do you mean tying the logit projection with the token embeddings? that was brought up in a previous issue and the flag is [here](https://github.com/lucidrains/nGPT-pytorch/blob/main/nGPT_pytorch/nGPT.py#L318)

**Comment by lucidrains** [comment]
@rotem154154 is weight tying even used that much in actual large language models being trained these days? i'm under the impression that it is not

**Comment by rotem154154** [comment]
Thank you! I obviously missed that XD. After setting the flag, the model ended up with 137k (0.1%) more parameters. That’s interesting—you’re probably right that it’s not used much. I’ll try both with and without it in my experiments to improve NanoGPT training, like here:
https://x.com/kellerjordan0/status/1845865698532450646

I’ll let you know if nGPT turns out to be successful!

**Comment by lucidrains** [comment]
@rotem154154 yes please do

and also if you hit some negative result (some unreported metric is bad) also let me know, through email or otherwise

thank you

**Comment by lucidrains** [comment]
from the claims in this paper, either one camp or the other will have egg on their face, so to speak.

**Comment by ghgmail** [comment]
Why is there two files: nGPT and nTransformers？  and what is the difference？ tks~！ @lucidrains @rotem154154
